### PR TITLE
Rewrite test in nock

### PR DESCRIPTION
**Pull request by xingoxu**
I would like to use [nock](https://github.com/nock/nock) to replace the test server.
- We can't verify the host/prefix from the simple server since we can only get `localhost`.
- I tried to use mock function to intercept the input & output and do unit test. But we can't test class constructor.

For above reasons, I rewrite the client, http tests in nock.

We still need test server for testing middleware, since nock can't use middleware directly.

### CODE REVIEW

**Comment by xingoxu** [comment]
Should be merged after #178 & #180

File: package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
   ],
   "scripts": {
     "pretest": "npm run build",
-    "test": "API_BASE_URL=http://localhost:1234/ TEST_PORT=1234 TS_NODE_CACHE=0 nyc mocha",
+    "test": "TEST_PORT=1234 TS_NODE_CACHE=0 nyc mocha",
```
Comment (ianwith) [inline package.json]: TEST_PORT is not used in the code

File: test/utils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { ensureJSON } from "../lib/utils";
+import { JSONParseError } from "../lib/exceptions";
+import { equal, ok } from "assert";
+
+describe("utils", () => {
+  it("fails when it's not an object", () => {
```
Comment (ianwith) [inline test/utils.spec.ts]: only `ensureJSON` is tested.

File: test/http.spec.ts [hunk 3]
```diff
@@ -2,56 +2,80 @@ import { deepEqual, equal, ok } from "assert";
...
   });
 
-  before(() => listen(TEST_PORT));
-  after(() => close());
+  before(() => nock.disableNetConnect());
+  afterEach(() => nock.cleanAll());
+  after(() => nock.enableNetConnect());
+
+  const interceptionOption = {
+    reqheaders: {
```
Comment (ianwith) [inline test/http.spec.ts]: `headers` might be better. Or `reqHeaders`

File: test/http.spec.ts [hunk 4]
```diff
@@ -61,106 +85,125 @@ describe("http", () => {
...
+      },
+    })
+      .post("/post/binary", buffer)
+      .reply(200, {});
+  };
+
   it("postBinary", async () => {
     const filepath = join(__dirname, "/helpers/line-icon.png");
     const buffer = readFileSync(filepath);
+    let scope = mockPostBinary(buffer, {
```
Comment (ianwith) [inline test/http.spec.ts]: use `const`.

File: test/http.spec.ts [hunk 5]
```diff
@@ -61,106 +85,125 @@ describe("http", () => {
...
-    equal(req.headers["user-agent"], `${pkg.name}/${pkg.version}`);
-    equal(req.headers["content-type"], "image/png");
-    equal(req.headers["content-length"], buffer.length);
+    equal(scope.isDone(), true);
   });
 
   it("postBinary with specific content type", async () => {
     const filepath = join(__dirname, "/helpers/line-icon.png");
     const buffer = readFileSync(filepath);
+    let scope = mockPostBinary(buffer, {
```
Comment (ianwith) [inline test/http.spec.ts]: Please check all these `let`

File: test/http.spec.ts [hunk 6]
```diff
@@ -2,56 +2,80 @@ import { deepEqual, equal, ok } from "assert";
...
+  "test-header-key": "Test-Header-Value",
+};
 
 describe("http", () => {
   const http = new HTTPClient({
-    baseURL: `http://localhost:${TEST_PORT}`,
-    defaultHeaders: {
-      "test-header-key": "Test-Header-Value",
-    },
+    baseURL,
```
Comment (ianwith) [inline test/http.spec.ts]:
```
new HTTPClient({
      baseURL
})
```
Since we don't need this param any more. Maybe it's better to remove `baseURL` and make tests based on different endpoints.

File: test/utils.spec.ts (hunk 2 above)
Comment (xingoxu) [reply test/utils.spec.ts]: only `ensureJSON` is uncovered so I only test the failed case.

File: package.json (hunk 1 above)
Comment (xingoxu) [reply package.json]: `TEST_PORT` is still used in middleware test.

File: test/utils.spec.ts (hunk 2 above)
Comment (ianwith) [reply test/utils.spec.ts]: so maybe change to `ensureJSON fails ...` will be better?

File: test/http.spec.ts (hunk 6 above)
Comment (xingoxu) [reply test/http.spec.ts]: Alright, but maybe it should be a new pull request?

**Comment by xingoxu** [comment]
Please review again & check #180 first. @ianwith

File: test/helpers/test-server.ts [hunk 7]
```diff
@@ -115,7 +67,7 @@ function listen(port: number, middleware?: express.RequestHandler) {
         res.status(400).send(err.raw);
         return;
       }
-      // https://github.com/expressjs/express/blob/master/lib/application.js#L162
+      // https://github.com/expressjs/express/blob/2df1ad26a58bf51228d7600df0d62ed17a90ff71/lib/application.js#L162
```
Comment (ianwith) [inline test/helpers/test-server.ts]: why is this changed?

Comment (xingoxu) [reply test/helpers/test-server.ts]: I just want to lock that commit so if master changed, we can know why we add this part at that time.
